scripts/testing/database-emulator-test.js

```diff
@@ -0,0 +1,121 @@
+ * limitations under the License.
+ */
+
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const { spawn } = require('child-process-promise');
+
+const EMULATOR_LINK =
+  'https://firebase.google.com/docs/database/downloads/database-emulator.jar';
+const EMULATOR_DIRECTORY = path.join(os.tmpdir(), 'firebase-emulators');
+const EMULATOR_FILENAME = 'database-emulator.jar';
+
+const EMULATOR_PORT = 8088;
+const EMULATOR_NAMESPACE = 'test-emulator';
+
+async function delay(seconds) {
+  return new Promise(resolve => setTimeout(resolve, 1000 * seconds));
+}
+
+async function downloadEmulator() {
+  return new Promise((resolve, reject) => {
+    console.log(`Creating directory [${EMULATOR_DIRECTORY}] ...`);
+    fs.mkdir(EMULATOR_DIRECTORY, err => {
+      if (err && err.code !== 'EEXIST') reject(err);
+      console.log(`Downloading emulator binary from [${EMULATOR_LINK}] ...`);
+      // const promise = spawn('firebase', ['setup:emulators:database']);
+      const promise = spawn('curl', [EMULATOR_LINK, '-o', EMULATOR_FILENAME], {
+        cwd: EMULATOR_DIRECTORY,
+        stdio: 'inherit'
+      });
+      resolve(promise);
+    });
+  });
+}
+
+async function launchEmulator(locked) {
+  const emulator = new Promise((resolve, reject) => {
+    // const promise = spawn('firebase', ['serve', '--only', 'database']);
+    const promise = spawn(
+      'java',
+      ['-jar', EMULATOR_FILENAME, '--port', EMULATOR_PORT],
+      { cwd: EMULATOR_DIRECTORY }
+    );
+    promise.catch(reject);
+    const process = promise.childProcess;
+    console.log(`Bringing up emulator, pid: [${process.pid}] ...`);
+    process.stdout.on('data', data => {
+      console.log(`[emulator-logs]: ${data}`);
+      // Working on better ways to signal readiness.
+      if (data.includes(`Listening on port ${EMULATOR_PORT}`)) {
+        setTimeout(() => resolve(process), 5000);
+      }
+    });
+  });
+  await emulator;
+  if (!locked) {
+    console.log('Loading rule {".read": true, ".write": true} to emulator ...');
+    await spawn(
+      'curl',
+      [
+        '-H',
+        'Authorization: Bearer owner',
+        '-X',
+        'PUT',
+        '-d',
+        '{"rules": {".read": true, ".write": true}}',
+        `http://localhost:${EMULATOR_PORT}/.settings/rules.json?ns=${EMULATOR_NAMESPACE}`
+      ],
+      { stdio: 'inherit' }
+    );
+  }
+  console.log('\n\nEmulator has started up successfully!\n\n');
+  return emulator;
+}
+
+async function runTest() {
+  const options = {
+    cwd: path.resolve(__dirname, '../../packages/database'),
+    env: Object.assign({}, process.env, {
+      USE_RTDB_EMULATOR: true,
+      RTDB_EMULATOR_PORT: EMULATOR_PORT,
+      RTDB_EMULATOR_NAMESPACE: EMULATOR_NAMESPACE
+    }),
+    stdio: 'inherit'
+  };
+  return Promise.all([
+    spawn('yarn', ['test:node'], options),
+    spawn('yarn', ['test:browser'], options)
+  ]);
+}
+
+(async () => {
```

Comment:
Hrm...  This looks a bit strange.

```
(async () => ...)();
```

That's going to define a lambda that returns a Promise, and then call the lambda, generating a Promise. But then it doesn't do anything with the returned Promise.  I guess it _does_ let you use await within the lambda code in order to sequence the actions.  But it's still a bit strange. I'd maybe extract the lambda into an actual named async function called `start()` or something. And then either handle the promise, or add a comment saying we don't need to.

Reply:
Done? The problem is that I need the reference to the emulator in the finally block. That's making the code not so streamlined.